astropy/cosmology/io.py

```diff
@@ -0,0 +1,351 @@
+    column "name") with a Mapping value will also be taken as metadata. If both
+    exist, they will be merged, with the column taking higher priority.
+
+    Many formats do not allow unit information, so make sure the values are in
+    the correct units for initializing the cosmology. For this reason ECSV
+    is the recommended file format for Cosmologies.
+
+    Examples
+    --------
+    TODO
+
+    Raises
+    ------
+    TypeError
+        If a parameter is not in the init signature and already has a value in
+        the metadata.
+    """
+
+    def __init__(self, instance, cls):
+        super().__init__(instance, cls, "read")
+
+    def __call__(self, *args, index=None, **kwargs):
+        # shortcuts for determining format
+        format = kwargs.get("format", None)
+        fmt = str(format).lower()
+        fend = str(args[0]).lower().endswith
+
+        # need to determine the format: is it JSON or ECSV?
+        # JSON?
+        if (fmt == "json") or (format is None and fend(".json")):
+            cosmos = self._read_json(*args, key=index, **kwargs)
+        # ECSV?
+        elif (fmt == "ascii.ecsv") or (format is None and fend(".ecsv")):
+            table = QTable.read(*args, **kwargs)
+            cosmos = self.from_table(table, index=index)
+        else:
+            raise ValueError(
+                "`format` must be JSON or 'ascii.ecsv' or None. "
+                "If format is None, `file` must have suffix '.json' or '.ecsv'."
+            )
+
+        return cosmos
+
+    def _read_json(self, file, key=None, **kwargs):
+        # read file, from path-like or file-like
+        if isinstance(file, (str, bytes, os.PathLike)):  # pathlike
+            with open(file, "r") as read_file:
+                data = read_file.read()
+        else:  # file-like : this also handles errors in dumping
+            data = file.read()
+
+        # parse file
+        mapping = json.loads(data)
+
+        return self.from_mapping(mapping, key=key)
+
+    def from_mapping(self, mapping, key=None):
+        """Load `~astropy.cosmology.Cosmology` from mapping object.
+
+        Parameters
+        ----------
+        mapping : mapping
+        key : hashable or None, optional
+            Any valid key into `mapping`.
+            If not None, this is used to select a cosmology if mapping is a
+            set of cosmologies.
+
+        Returns
+        -------
+        `~astropy.cosmology.Cosmology`
+
+        """
+        # get params subset from mpping
+        params = mapping[key] if key is not None else mapping
+
+        # determine class
+        subclasses = {c.__qualname__: c for c in
+                      self._cls.__subclasses__(deep=True)}
+        cls_name = params.pop("cosmology", self._cls.__qualname__)
+        try:
+            cosmo_cls = subclasses[cls_name]
+        except KeyError:
+            raise KeyError(f"{cls_name} is not a subclass of {cls.__qualname__}.")
+
+        # match keys to arguments
```

Comment:
Agreed. I'll simplify it. It was motivated so that a mapping like ``{"cosmology": "LambdaCDM", "flat": True}`` could initialize a FlatLambdaCDM cosmology. With ``cosmo_cls(**mapping)`` it will error because flat is not an accepted parameter. If this proves problematic for a user, then we can figure out a more general solution.

Reply:
Turns out I kind of needed the more general solution in ``realizations.py`` for the parameters in ``parameters.py`` and the eventual option for user parameters. So I refactored and figured out a much simpler way to do argument matching using ``BoundArguments`` rather than going through the signature directly.

Reply:
Superseded by subsequent discussions. Resolving here.